### DISCUSSION TITLE
CRUD tutorial for absolute beginners?

### DISCUSSION THREAD

**Discussion by theArianit** [TypedProject/ts-express-decorators#1678] [Q&A]
Hello beautiful people,
as the title says, is there any CRUD tutorial/project to get started with ts.ed? for me these kind of projects/tutorials are the best to get started into the most basics. Or any tutorial for absolute beginners would be most welcome.
Thank you for any help.

**Comment by Romakita** [comment]
Hello @theArianit 

The tutorial doesn't really exists right know, but it's planned!. I'm really sorry about that (and also for my late answers).
You can find video and topics on google (example: https://www.youtube.com/watch?v=Flckoks7yZE)
I'll try to works on this content ASAP

Romain

**Reply by theArianit** [reply]
Hello @Romakita 
thank you for your reply. Nice to hear that it is planned. Can't wait for it to be ready.
Kind regards.

**Reply by Romakita** [reply]
Ok I understand!

The first step is to generate a new project with the Ts.ED CLI:

```typescript
tsed init .
```

You can select any options like a BDD with typeORM/mongoose etc...

When the project is generated, you can start the server:

```
npm run start
```

**Reply by Romakita** [reply]
## Create a controler

To do that, use the CLI:

```
tsed g controllers users
```

You should have something like that:

```typescript
@Controller("/users")
export class UsersController {
  @Get("/")
  get() {
     return []
  }
}
```

**Reply by Romakita** [reply]
## Create our first CRUD Controller

Ts.ED provide many decorator to create routes based on the different verbs (GET, POST, PUT, DELETE). See https://tsed.io/docs/controllers.html

Edit the UsersController and add the following methods:

```typescript
import {Controller} from "@tsed/common";
mport {PathParams} from tsed/platform-params
import {Get, Post, Put, Returns} from "@tsed/schema";
import {NotFound} from "@tsed/exceptions";

@Controller("/users")
export class UsersController {
  private users: any[] = [{id: '1', name: 'John'}];
  
  @Get("/:id")
  getById(@PathParams("id") id: string) {
     const user =  this.users.find(user => user.id === id);
     
     if (!user) {
        throw new NotFound("User not found");
     }
     
     return user;
  }
  
  @Post("/")
  @Returns(201)
  create(@BodyParams() user: any) {
     const user = {...user, id: this.users.length + 1}
     this.users.push(user);
     
     return user;
  }
  
  @Put("/:id")
  update(@PathParams('id') id: string; @BodyParams() user: any) {
     const index =  this.users.findIndex(user => user.id === id);
     
     if (index === -1) {
        throw new NotFound("User not found");
     }

     this.users[index] = user;
     
     return user;
  }
  
  @Delete("/:id")
  @Returns(204)
  delete() {
     const item =  this.users.find(user => user.id === id);
     
     if (!item) {
        throw new NotFound("User not found");
     }
     
     this.users = this.users.filter(user => user.id === id);
  }
  
  @Get("/")
  getAll() {
     return this.users;
  }
}
```

**Reply by Romakita** [reply]
Tell me if you have any error @theArianit

If all is ok, we'll move on the next step ;)

**Reply by theArianit** [reply]
Thank you for the reply. I am currently working on some family issues but I will take a proper look after next week.
There is something I stumbled upon, and forgive me for my ignorance on this part, but I have read somewhere that mongoose doesn't quiet like typescript and therefore there is this [typegoose](https://typegoose.github.io/typegoose/) library. Again, excuse me for my ignorance on this part, but what has to be done to support a library in a javascript/typescript framework, more specific tsed, which isn't mentioned in the official documentation. 
thank you again for your assistance and patient.

**Reply by Romakita** [reply]
Becauce typegoose wasn’t correctly design 3 years ago. Consumes metada from typegoose was a real pain point for me. The benefice wasn’t really interesting rather than the probleme to use with another framework. 

Today, the tsed mongoose version is mature, it consume the same metada to generate the swagger spec/jsonschema and the mongoose schema. 
also @tsed/mongoose can be used in standalone (you just need to have @tsed/di and @tsed/schema).

see you ^^

**Reply by theArianit** [reply]
Hello. I need to make an extra question to this thread. What is the easiest and best practice way to work with users, login/logout, user session, authentication, etc. I am reading the documentation, but coming from an ASP.Net background, all those things seem so complicated on the first glance (maybe a second reading will clarify things). Therefore a simplification with practical examples would be most welcome. What interest me more is that in .Net you can just use everything what asp.net offers you out of the box to create users, and it is very easy to create roles and authentication. All you have to do is small tweaks on what already exists, and those are relative easy to find (password rules, hashing and salting of passwords, etc.). does tsed have something similar to that? Every tips and tricks is welcome. Thank you

**Reply by Romakita** [reply]
Hello @theArianit 

You are confusing framework with environment. Ts.ED is a framework with limited scope and ASP.Net is an environment with many
tools provided by Microsoft (Language, LDAP, Roles...). 

By nature, Node.js provide a minimal list of feature. NPM extends the possibilities by installing additional package like Ts.ED. 

Ts.ED scope is to provide a framework to write  your **Rest Application** on top of Express/Koa (controller/service/etc...) with a minimal code. But it doesn't provide out of box a feature to manage user (I haven't the resource for that 💰). It's a specific for each project. NPM have a lot of modules that can be used with Ts.ED without extra configuration. Actually, Ts.ED recommend three solutions:

- Passport.js, which is a good solution to manage security with any auth third parties like Facebook/google/oauth or with a local database (combined with https://adminjs.co/ it can be a good deal)
- OIDC provider which is an official OIDC implementation in node.js but there is no role management out of the box,
- Keycloak, it's another OIDC provider not implemented in node.js (java)

But, You can find any other solution on NPM and integrate it with your Ts.ED project.

I hope my answer will help you
Romain

**Reply by theArianit** [reply]
Hello @Romakita 
I have to thank you for your time you invest to answer my noob question, even though I have some years of experience in programming. As for now I see that I need to change the way I approach nodejs/tsed compared to Asp.Net (full environment vs framework). I will take a look into Passport.js, since it looks more promising from the options you gave in your answer. I will try to create a project to manage users, to begin with tsed, and to maybe switch from asp.net to a nodejs framework(tsed) to work on my private projects.

Kind regards.

**Reply by theArianit** [reply]
I finally got the time to get started. I tried your users controller. I had to make 2 small fixes. You forgot the id parameter in the get decorator of getUser and also your forgot to import PathParams from tsed/platform-params. And it works :)

**Reply by Romakita** [reply]
The next step is to create a service that consume a database :)
Adding all logic in a controller is a bad pattern, it's preferable to create a Service that contains the business rules and Repository to consume the database. In our previous example, there is no business rule so we'll stay focused on creating a Repository that store and consume the data:

```typescript
import {NotFound} from "@tsed/exceptions";
import {Injectable} from "@tsed/common";

@Injectable()
export class UsersRepository {
  private users: any[] = [{id: '1', name: 'John'}];
  
  getById(id: string) {
     const user = this.users.find(user => user.id === id);
     
     if (!user) {
        throw new NotFound("User not found");
     }
     
     return user;
  }

  create(user: any) {
     const user = {...user, id: this.users.length + 1}
     this.users.push(user);
     
     return user;
  }
  
  update(user: any) {
     const index =  this.users.findIndex(u => u.id === user.id);
     
     if (index === -1) {
        throw new NotFound("User not found");
     }

     this.users[index] = user;
     
     return user;
  }

  delete() {
     const item =  this.users.find(user => user.id === id);
     
     if (!item) {
        throw new NotFound("User not found");
     }
     
     this.users = this.users.filter(user => user.id === id);
  }
  
  getAll() {
     return this.users;
  }
}
```

And the controller:

```typescript
import {Controller} from "@tsed/common";
import {PathParams} from tsed/platform-params
import {Get, Post, Put, Returns} from "@tsed/schema";
import {UsersRepository} from "../services/UsersRepository";

@Controller("/users")
export class UsersController {
  @Inject()
  repository: UsersRepository;
  
  @Get("/:id")
  getById(@PathParams("id") id: string) {
      return this.repository.getById(id);
  }
  
  @Post("/")
  @Returns(201)
  create(@BodyParams() user: any) {
     return this.repository.create(user);
  }
  
  @Put("/:id")
  update(@PathParams('id') id: string; @BodyParams() user: any) {
     return return this.repository.update({...user, id});
  }
  
  @Delete("/:id")
  @Returns(204)
  delete(@PathParams("id") id: string) {
     return return this.repository.delete(id);
  }
  
  @Get("/")
  getAll() {
     return this.repository.getAll();
  }
}
```

**Reply by theArianit** [reply]
I understand. This is how I do it in asp.net. I will update you when I follow up.

**Reply by theArianit** [reply]
Can you help me fix the update method? Because user.id is undefined and therefore the index -1. I can't figure out the problem.

**Reply by Romakita** [reply]
```typescript
  @Put("/:id")
  update(@PathParams('id') id: string; @BodyParams() user: any) {
     return return this.repository.update({...user, id});
  }
```

**Reply by theArianit** [reply]
thanks. what's the next thing you suggest? I would like to know a bit more about interface vs/and model, because these two confuse me in nodejs apps.

**Reply by Romakita** [reply]
Interface have no consistency in javascript. It's only for the type checking. 

Models are based on the classe Javascript. Using decorators, you can describe a model with validation rules (based on JsonSchema).
Ts.ED uses the model to generate the Swagger aka OpenSpec for your different routes and it also use the model to validate the input before mapping input data to the model.

Read this page: https://tsed.io/docs/model.html
And try to declare a User model and use it in the controller and service. If you have installed the `@tsed/swagger`, you'll the documentation change and reflect you code implementation.


See you
romain
